app/src/main/java/fr/free/nrw/commons/customselector/ui/adapter/ImageAdapter.kt

```diff
@@ -87,17 +124,81 @@ class ImageAdapter(
                 notifyItemRangeChanged(updatedPosition, images.size)
             }
         } else {
-            val selectedIndex = ImageHelper.getIndex(selectedImages, image)
+            val sharedPreferences: SharedPreferences =
+                context.getSharedPreferences(CUSTOM_SELECTOR_PREFERENCE_KEY, 0)
+            val switchState =
+                sharedPreferences.getBoolean(SWITCH_STATE_PREFERENCE_KEY, true)
+            val selectedIndex: Int = if(switchState) {
+                ImageHelper.getIndex(selectedImages, image)
+            } else {
+                if(mapActionableImages.size > position) {
+                    ImageHelper
+                        .getIndex(selectedImages, ArrayList(mapActionableImages.values)[position])
+                } else {
+                    -1
+                }
+            }
             val isSelected = selectedIndex != -1
             if (isSelected) {
-                holder.itemSelected(selectedIndex + 1)
+                holder.itemSelected(selectedImages.size)
             } else {
-                holder.itemUnselected();
+                holder.itemUnselected()
             }
-            Glide.with(holder.image).load(image.uri).thumbnail(0.3f).into(holder.image)
-            imageLoader.queryAndSetView(holder, image)
+
+            scope.launch {
+                imageLoader.queryAndSetView(
+                        holder, image, ioDispatcher, defaultDispatcher
+                )
+                val sharedPreferences: SharedPreferences =
+                    context.getSharedPreferences(CUSTOM_SELECTOR_PREFERENCE_KEY, 0)
+                val switchState =
+                    sharedPreferences.getBoolean(SWITCH_STATE_PREFERENCE_KEY, true)
+                if (!switchState) {
```

Comment:
Please add comments to the case statements, as we have many nested statements it would be very helpful. 
You can add it to the task lists just that you don't forget and come back here after important stuff is done.

Reply:
@4D17Y4 Should I add comments before every conditional statement or just create a TODO list where I will mention the logic?

Reply:
I think it is better to just write all the comments now (before each conditional statement). In general I think your code needs much more comments explaining the logic, too. Thanks! 🙂